jupyter_server/serverapp.py

```diff
@@ -1487,31 +1490,96 @@ def init_server_extension_config(self):
+            if enabled:
+                # Search for server extensions by loading each extension module
+                # and looking for the `_jupyter_server_extension_paths` function.
+                try:
+                    mod, metadata_list = _get_server_extension_metadata(module_name)
+                    for metadata in metadata_list:
+                        # Check if this server extension is a ExtensionApp object.
+                        extapp = metadata.get('app', None)
+                        extloc = metadata.get('module', None)
+                        # Load the extension module.
+                        if extapp and extloc:
+                            if not issubclass(extapp, JupyterApp):
+                                raise TypeError(extapp.__name__ + " must be a subclass of JupyterApp.")
+                            # Initialize extension app
+                            app = extapp(parent=self)
+                            # Update the app's config, setting
+                            # any traits given by the serverapp's
+                            # parsed command line args.
+                            app.update_config(app.config)
+                            # Load any config from an extension's
+                            # config file and make update the
+                            # app's config again.
+                            app.load_config_file()
+                            # Pass any relevant config to the
+                            # serverapp's (parent) config.
+                            self.update_config(app.config)
+                            self.enabled_extensions[app.extension_name] = app
+                        elif extloc:
+                            extmod = importlib.import_module(extloc)
+                            func = getattr(extmod, 'load_jupyter_server_extension', None)
+                            if func is None:
+                                log_msg = _(
+                                    "{module_name} is enabled but no "
+                                    "`load_jupyter_server_extension` function "
+                                    "was found.".format(module_name=module_name)
+                                )
+                                self.log.warning(log_msg)
+                            else:
+                                self.enabled_extensions[extloc] = extmod
+                        else:
+                            log_msg = _(
+                                "{module_name} is missing critical metadata. Check "
+                                "that _jupyter_server_extension_paths returns `app` "
+                                "and/or `module` as keys".format(module_name=module_name)
+                            )
+                            self.log.warn(log_msg)
+                except KeyError:
+                    # A KeyError suggests that the module does not have a
+                    # _jupyter_server_extension-path.
```

Comment:
Would it be cleaner to put this somewhere closer to the actual lines that raise this exception to ensure we don't misattribute other `KeyError`s?

Reply:
:+1: refactored slightly to fix this. Thanks!